### PR TITLE
Compatability with 0.36.1

### PR DESCRIPTION
**Pull request by kalinon**
This will be needed for compatibility for 0.34.0. However some kemal/sessions will need to be updated before this cspec can be fixed.

### CODE REVIEW

File: src/sidekiq/logger.cr [hunk 1]
```diff
@@ -1,57 +1,82 @@
...
+      name = target.class.name.underscore.gsub("::", ".")
+      ::Log.builder.bind(name, :info, Sidekiq::Logger::PrettyBackend.new(log_target))
+      ::Log.for(name)
     end
   end
 end
 
 class Fiber
-  property logging_context : Array(String)?
+  property sidkiq_logging_context : Array(String)?
```
Comment (mperham) [inline src/sidekiq/logger.cr:81]: misspelled

File: shard.yml [hunk 2]
```diff
@@ -20,7 +20,7 @@ dependencies:
     branch: master
   kemal-session:
     github: kemalcr/kemal-session
-    commit: a16f164d3dcfc6e720d566f9085285abdd5b7825
+    version: ~> 0.10.0
```
Comment (mperham) [inline shard.yml:23]: This should be a separate commit or PR, right?

File: src/sidekiq/logger.cr (hunk 1 above)
Comment (kalinon) [reply src/sidekiq/logger.cr]: Good catch. Fixed.

File: shard.yml (hunk 2 above)
Comment (kalinon) [reply shard.yml]: Can be, ive reverted it.

Comment (mperham) [reply shard.yml]: Add it as a separate commit here so the build passes and I can merge.

Comment (kalinon) [reply shard.yml]: It wont pass, because even the newest version of `kemalcr/kemal-session` is broken on 0.34.0. We cant update the logger before 0.34.0 (because `Log` doesnt exist yet). So we need to wait for a 0.34.0 version of `kemalcr/kemal-session`. I will try to make a PR there as well.

File: src/sidekiq/server/exception_handler.cr [hunk 3]
```diff
@@ -1,19 +1,19 @@
...
 
       def initialize(@output)
       end
 
       def call(ex : Exception, ctxHash : Hash(String, JSON::Any)? = nil)
-        @output.warn(ctxHash.to_json) if ctxHash && !ctxHash.empty?
-        @output.warn "#{ex.class.name}: #{ex.message}"
-        @output.warn ex.backtrace.join("\n")
+        @output.warn { ctxHash.to_json } if ctxHash && !ctxHash.empty?
+        @output.warn(exception: ex) { "#{ex.class.name}: #{ex.message}" }
```
Comment (Sija) [inline src/sidekiq/server/exception_handler.cr:15]: Redundant

```suggestion
        @output.warn(exception: ex) { }
```

File: src/sidekiq/server/processor.cr [hunk 4]
```diff
@@ -82,9 +82,9 @@ module Sidekiq
     def handle_fetch_exception(ex)
       if !@down
         @down = Time.local
-        @mgr.logger.error("Error fetching job: #{ex}")
+        @mgr.logger.error { "Error fetching job: #{ex}" }
         ex.backtrace.each do |bt|
-          @mgr.logger.error(bt)
+          @mgr.logger.error { bt }
         end
```
[the remark this answers is not in the document]
Comment (kalinon) [reply src/sidekiq/server/processor.cr:88]: probably. I did mostly a find and replace with a regex on these spots

File: shard.yml [hunk 5]
```diff
@@ -17,13 +17,16 @@ dependencies:
...
   kemal:
     github: kemalcr/kemal
-    branch: master
+    # branch: master
+    version: ~> 0.26.0
   kemal-session:
     github: kemalcr/kemal-session
-    commit: a16f164d3dcfc6e720d566f9085285abdd5b7825
+    version: ">= 0.12.0"
+    # commit: a16f164d3dcfc6e720d566f9085285abdd5b7825
```
Comment (Sija) [inline shard.yml:25]:
```suggestion
```

File: shard.yml [hunk 6]
```diff
@@ -17,13 +17,16 @@ dependencies:
...
   kemal-csrf:
     github: kemalcr/kemal-csrf
-    version: 0.5.0
+    version: ">= 0.5.0"
+    # version: 0.5.0
```
Comment (Sija) [inline shard.yml:29]:
```suggestion
```

File: shard.yml [hunk 7]
```diff
@@ -17,13 +17,14 @@ dependencies:
     version: ~> 0.2.3
   kemal:
     github: kemalcr/kemal
-    branch: master
+    # branch: master
```
Comment (Sija) [inline shard.yml:20]:
```suggestion
```

File: src/sidekiq/server/exception_handler.cr [hunk 8]
```diff
@@ -22,9 +21,8 @@ module Sidekiq
         begin
           handler.call(ex, ctxHash)
         rescue ex2
-          ctx.logger.error "!!! ERROR HANDLER THREW AN ERROR !!!"
-          ctx.logger.error ex2
-          ctx.logger.error ex2.backtrace.join("\n")
+          ctx.logger.error { "!!! ERROR HANDLER THREW AN ERROR !!!" }
+          ctx.logger.error(exception: ex2) { }
```
Comment (Sija) [inline src/sidekiq/server/exception_handler.cr:25]:
```suggestion
          ctx.logger.error(exception: ex2) { "!!! ERROR HANDLER THREW AN ERROR !!!" }
```

**Comment by kalinon** [comment]
Ive added some more commits to this PR to be compatible with building with `preview_mt`. I subbed out the pooling code to use `DB::Pool` which is thread safe.

**Comment by n-rodriguez** [comment]
Hi there! Any news?

File: shard.yml [hunk 9]
```diff
@@ -17,14 +17,21 @@ dependencies:
...
+    # version: ">= 0.26.0"
+    commit: a8c0f09b858162bd13c96663febef5527b322a32
```
Comment (Sija) [inline shard.yml:22]:
```suggestion
    version: "~> 0.27.0"
```

File: src/sidekiq/server/cli.cr [hunk 10]
```diff
@@ -5,15 +5,15 @@ require "../server"
 
 module Sidekiq
   class CLI
-    getter logger : ::Logger
+    getter logger : ::Log = ::Log.for(Sidekiq::CLI)
```
Comment (Sija) [inline src/sidekiq/server/cli.cr:8]:
```suggestion
    getter logger : ::Log = ::Log.for(self)
```

File: src/sidekiq/pool.cr [hunk 11]
```diff
@@ -62,9 +64,12 @@ module Sidekiq
     end
 
     def initialize(redis_cfg : RedisConfig)
-      @pool = ConnectionPool(Redis).new(capacity: redis_cfg.pool_size, timeout: redis_cfg.pool_timeout) do
+      @pool = DB::Pool.new(
+        checkout_timeout: redis_cfg.pool_timeout,
+        max_pool_size: redis_cfg.pool_size
+      ) {
```
Comment (Sija) [inline src/sidekiq/pool.cr:70]:
```suggestion
      ) do
```

File: src/sidekiq/pool.cr [hunk 12]
```diff
@@ -62,9 +64,12 @@ module Sidekiq
...
         redis_cfg.new_client
-      end
+      }
```
Comment (Sija) [inline src/sidekiq/pool.cr:72]:
```suggestion
      end
```

File: src/sidekiq/pool.cr [hunk 13]
```diff
@@ -84,9 +89,20 @@ module Sidekiq
     #     end
     #
     def redis
-      @pool.connection do |conn|
+      conn = @pool.checkout
+      begin
         yield conn
+      ensure
+        @pool.release conn
```
Comment (Sija) [inline src/sidekiq/pool.cr:96]:
```suggestion
        @pool.release(conn)
```

File: spec/spec_helper.cr [hunk 14]
```diff
@@ -1,19 +1,20 @@
 require "spec"
+require "json_mapping"
```
[the remark this answers is not in the document]
Comment (kalinon) [reply spec/spec_helper.cr:2]: will do

**Comment by kalinon** [comment]
Pulled in suggestions from @Sija and @n-rodriguez (Thanks!). Updated deps and code to work with `0.36.1`

File: spec/spec_helper.cr (hunk 14 above)
[the remark this answers is not in the document]
Comment (Sija) [reply spec/spec_helper.cr:2]: Still there ;)

File: src/sidekiq/logger.cr [hunk 15]
```diff
@@ -1,57 +1,73 @@
...
-      Fiber.current.object_id.to_s(36, io)
-      io << " "
-      io << Sidekiq::Logger.context
-      io << " "
-      io << severity
-      io << ": "
-      io << message
+
+    class PrettyBackend < ::Log::IOBackend
+      def initialize(@io = STDOUT)
```
[the remark this answers is not in the document]
Comment (n-rodriguez) [reply src/sidekiq/logger.cr:42]: still missing

**Review by n-rodriguez — CHANGES_REQUESTED** [review]

File: src/sidekiq/server/cli.cr [hunk 16]
```diff
@@ -5,15 +5,15 @@ require "../server"
 
 module Sidekiq
   class CLI
-    getter logger : ::Logger
+    getter logger : ::Log = ::Log.for("sidekiq")
```
Comment (n-rodriguez) [inline src/sidekiq/server/cli.cr:8]: I think it should be `Sidekiq::Logger.build("sidekiq")`. `::Log.for("sidekiq")` bypasses some setup steps in `Sidekiq::Logger.build`. The result is an empty output when starting the worker.

**Review by n-rodriguez — CHANGES_REQUESTED** [review]

File: src/sidekiq/server/cli.cr (hunk 16 above)
Comment (n-rodriguez) [reply src/sidekiq/server/cli.cr:8]: btw, awesome work :+1:

File: src/sidekiq/server/cli.cr [hunk 17]
```diff
@@ -5,15 +5,15 @@ require "../server"
 
 module Sidekiq
   class CLI
-    getter logger : ::Logger
+    getter logger : ::Log = Sidekiq::Logger.build("sidekiq")
```
Comment (Sija) [inline src/sidekiq/server/cli.cr:8]:
```suggestion
    getter logger : ::Log
```

File: src/sidekiq/server/cli.cr [hunk 18]
```diff
@@ -5,15 +5,15 @@ require "../server"
...
 
     def initialize(args = ARGV)
       @concurrency = 25
       @queues = [] of String
       @timeout = 8
       @environment = "development"
       @tag = ""
-      @logger = Sidekiq::Logger.build
+      # @logger = Sidekiq::Logger.build(self)
```
Comment (Sija) [inline src/sidekiq/server/cli.cr:16]:
```suggestion
      @logger = Sidekiq::Logger.build(self)
```

File: src/sidekiq/server/cli.cr [hunk 19]
```diff
@@ -58,10 +58,10 @@ module Sidekiq
     def run(svr)
       # hack to avoid printing banner in test suite
       print_banner if logger == @logger
-      logger.info "Sidekiq v#{Sidekiq::VERSION} in Crystal #{Crystal::VERSION}"
-      logger.info Sidekiq::LICENSE
-      logger.info "Upgrade to Sidekiq Enterprise for more features and support: http://sidekiq.org"
-      logger.info "Starting processing with #{@concurrency} workers"
+      logger.info { "Sidekiq v#{Sidekiq::VERSION} in Crystal #{Crystal::VERSION}" }
+      logger.info { Sidekiq::LICENSE }
+      logger.info { "Upgrade to Sidekiq Enterprise for more features and support: http://sidekiq.org" }
```
Comment (Sija) [inline src/sidekiq/server/cli.cr:63]:
```suggestion
      logger.info { "Upgrade to Sidekiq Enterprise for more features and support: https://sidekiq.org" }
```

File: src/sidekiq/server/heartbeat.cr [hunk 20]
```diff
@@ -92,7 +92,7 @@ module Sidekiq
         ::Process.kill(Signal.parse(msgs), ::Process.pid)
       rescue e
```
Comment (Sija) [inline src/sidekiq/server/heartbeat.cr:93]:
```suggestion
      rescue ex
```

File: src/sidekiq/server/heartbeat.cr [hunk 21]
```diff
@@ -92,7 +92,7 @@ module Sidekiq
...
         # ignore all redis/network issues
-        svr.logger.error("heartbeat: #{e.message}")
+        svr.logger.error { "heartbeat: #{e.message}" }
```
Comment (Sija) [inline src/sidekiq/server/heartbeat.cr:95]:
```suggestion
        svr.logger.error(exception: ex) { "heartbeat" }
```

File: src/sidekiq/server/processor.cr [hunk 22]
```diff
@@ -60,13 +58,11 @@ module Sidekiq
...
     def get_one
-      begin
-        work = @mgr.fetcher.retrieve_work(@mgr)
-        (@mgr.logger.info { "Redis is online, #{Time.local - @down.not_nil!} sec downtime" }; @down = nil) if @down
-        work
-      rescue ex
-        handle_fetch_exception(ex)
-      end
+      work = @mgr.fetcher.retrieve_work(@mgr)
+      (@mgr.logger.info { "Redis is online, #{Time.local - @down.not_nil!} sec downtime" }; @down = nil) if @down
```
Comment (Sija) [inline src/sidekiq/server/processor.cr:62]:
```suggestion
      if @down
        @mgr.logger.info { "Redis is online, #{Time.local - @down.not_nil!} sec downtime" }
        @down = nil
      end
```

File: src/sidekiq/server/processor.cr [hunk 23]
```diff
@@ -82,10 +78,7 @@ module Sidekiq
     def handle_fetch_exception(ex)
       if !@down
         @down = Time.local
-        @mgr.logger.error("Error fetching job: #{ex}")
-        ex.backtrace.each do |bt|
-          @mgr.logger.error(bt)
-        end
+        @mgr.logger.error(exception: ex) { "Error fetching job: #{ex.message}" }
```
Comment (Sija) [inline src/sidekiq/server/processor.cr:81]:
```suggestion
        @mgr.logger.error(exception: ex) { "Error fetching job" }
```

File: src/sidekiq/server/server.cr [hunk 24]
```diff
@@ -14,15 +14,15 @@ module Sidekiq
...
+    getter logger : ::Log
     getter labels : Array(String)
     getter queues : Array(String)
     getter tag : String
     getter identity : String
     getter hostname : String
 
     def initialize(@environment = "development", @queues = ["default"],
-                   @concurrency = 25, @logger = Sidekiq::Logger.build)
+                   @concurrency = 25, @logger = Sidekiq::Logger.build(Sidekiq::Server))
```
Comment (Sija) [inline src/sidekiq/server/server.cr:25]:
```suggestion
                   @concurrency = 25, @logger = Sidekiq::Logger.build(self))
```

**Comment by hugopl** [comment]
I'm using this branch from @xaviablaza and it works.

**Comment by Sija** [comment]
Looks like @mperham doesn't give a damn.

**Comment by mperham** [comment]
Hi folks, I don't support this project anymore which means I ignore this repo and all comments **until** someone explicitly @'s me and politely asks me to merge. This is all it takes:

```
Hey @mperham, this is ready to go.
```

I would also be happy to see more maintainers; folks who want commit access so they can maintain this repo without me. If the Crystal community wants to fork the project and take over, that's fine -- I would just ask that the project name be changed. Sidekiq is me and I don't want others to be confused.

I also block people who add toxic commentary. Don't bring sarcasm, hostility or passive aggression into my repos. Thank you.

**Comment by hugopl** [comment]
@xaviablaza, would you mind in create a new PR with your changes? Since we don't know if the original author of this PR still on this.

I could create the PR too, but better have the original patch authors doing this.

IMO keeping the project here with non official support status and more maintainers that can approve/merge stuff bring more visibility than create a fork with a different name, so this would be better for both, Sidekiq and Crystal.

**Comment by xaviablaza** [comment]
I can do that this weekend.

**Comment by marcoroth** [comment]
@mperham maybe it would make sense to mention the current status of the project in the README or in a dedicated "looking for maintainers" issue.

**Comment by hugopl** [comment]
> I can do that this weekend.

I just noticed that the tests still failing, so if you don't mind I'll try to take a look to.

**Comment by hugopl** [comment]
Closing this, since https://github.com/mperham/sidekiq.cr/pull/92 got merged.
